### DISCUSSION TITLE
Fire Event when TDialog Window is closed on the Close-Button-Character

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#144] [Q&A] [answered]
Hello,
I have a TDialog, but I can not trace the "Close" Action, when I click the left top Corner, to close the TDialog.
I have add flags, but the handleEvent Event-Handler shows me not the MessageBox.

```
options   |= ofSelectable | ofTopSelect;
        flags     |= wfMove
                  |  wfGrow
                  |  wfClose
                  |  wfZoom;
...
void handleEvent(TEvent & event ){
    TWindow::handleEvent( event );
    if (event.what == evCommand) {
        switch (event.message.command) {
            case cmClose:
            messageBox("blaa blaa",mfInformation|mfOKButton);
                close();
                clearEvent(event);
                break;
        }
    }
}
```
How can I make Action's when Windows/Dialogs are closed ?

**Comment by magiblot** [comment, marked as answer]
That's because the `cmClose` command [is also handled by `TWindow::eventHandle`](https://github.com/magiblot/tvision/blob/0917f04206fcfa1973dc49547af74e7399241899/source/tvision/twindow.cpp#L117-L133). So, in this case, you should check for the `cmClose` command **before** calling `TWindow::eventHandle`.

However, you must take into account that [the `close()` method destroys the current object](https://github.com/magiblot/tvision/blob/0917f04206fcfa1973dc49547af74e7399241899/source/tvision/twindow.cpp#L64-L71), so you must make sure that you do not interact with it after calling `close()`: do not call `clearEvent` nor `TWindow::handleEvent` if you called `close()`. For example:

```c++
void handleEvent(TEvent & event ){
    // Handle cmClose first to override the handling in TWindow.
    if (event.what == evCommand && event.message.command == cmClose)
    {
        messageBox("blaa blaa",mfInformation|mfOKButton);
        clearEvent(event);
        close();
        // Don't do anything after close().
        return;
    }
    TWindow::handleEvent( event );
    // Handle other events...
}
```

**Comment by magiblot** [comment]
Or even better, override the `close` method:

```c++
void close() override {
    messageBox("blaa blaa",mfInformation|mfOKButton);
    TWindow::close();
}
```

Then you won't have to do anything special in your `handleEvent`.

**Comment by paule32** [comment]
Thank You.
This was a great Solution !
